cpp/src/umap/umap_api.cpp

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "umap_api.h"
+#include "umap.hpp"
+
+void initUmapParams(ML::UMAPParams *params) {
```

Comment:
C++ interface can afford default arguments. However, in case of C, we can either:

Case1: create a wrapper C-struct to have users pass the umap-params. We can expose a similar such method which can populate reasonable defaults.
```c
// declarations
typedef struct {
    /// all UMAP params for C-interface
} CumlUmapParams;
void initUmapParams(CumlUmapParams *p);

// Usage:
CumlUmapParams p;
initUmapParams(&p);
// user can now customize the params if for some of them, the default values are not what they want
cumlSpUmapFitSupervised(..., p);
```
This approach will need a separate internal method to convert `CumlUmapParams` to `ML::UMAPParams`.

Case2: expose all these params along the interface, which means users will have to pass those arguments always!

The first approach can be simpler for the users. Hence I prefer this approach. But it involves some work at our end to enable it. Inviting @dantegd @cjnolet and @muellren for discussion on this one.

Reply:
Case 1: The problem is that this requires the caller to allocate memory for the struct, thereby know the size of the struct, and the `offsetof` of the members that are to be changed from their default. For this reasons, this case would prevent me from using cuML in grCUDA. 

Case 2: That is why there are different functions!

Reply:
For case1, since you're talking about allocating memory, I presume you cannot create them on stack?

Also, why would caller need to know about the size of such a struct?

Reply:
Stack or heap does not matter. You need to pass a pointer to some object (struct). 
Obviously, `sizeof` does NOT work because the whole point is that the client does not have to know size and layout of the struct. IMHO there is not much point in providing a C API which essentially only provides a demangled symbol plus the same struct as the `UMAPParam` class. 

I would favor a construction function that would return a pointer (or handle) to a param object and a function to manipulate the individual parameters (`setUmapParameter(param_ptr, N_EPOCH, 500)`) but as far as I know cuML functions must not allocate memory (as part of the whole statelessness approach).

Reply:
I agree with @teju85 here that it's hard to justify an API for an algorithm that doesn't allow the setting of the hyperparameters for tuning the behavior of the algorithm. 

I would be in favor in allowing these parameters to be passed in as arguments, as we have done in other algorithms. I understand that might be a pain point for calling the function, though. Would documenting the reasonable default values here work? IMO, the struct-based approach @teju85 mentioned might make it easier on the user. 

In other C APIs I have worked with, passing structs around for complex arguments (and to maintain state across stateless functions like UMAP's fit() and predict() functions here) is just par for the course. To me it seems like the added complexity is necessary here to be able to reap the portability benefits of C over just using the C++ API.

Reply:
Documentation of default arguments is insufficient in my opinion as it means that every use of the function would essentially needs to have a copy of the default parameters. Furthermore, if we find to change the default values (in the documentation) they would have to manually updated at every usage site. 

The coupling in the API should be as loose as possible. Exposing a user-defined type like the struct simply makes the API harder to use for non-statically linked simple C programs. Again, if you insist on passing around structs, the API should have an allocator+constructor function that returns a pointer to a newly allocated and default-initialized struct PLUS a getter and a setter function for the attributes.

I would favour option 1.

**Option 1:**
```C
typedef enum {
  UMAP_PARAM_N_NEIGHBORS = 1,
  UMAP_PARAM_N_EPOCHS = 2,
  ...
} UmapParamAttr;

struct UmapParams;
typedef struct UmapParams UmapParams_t;

UmapParams_t *createDefaultUmapParams();
void freeUmapParams(UmapParams_t *params);
void setUmapParamAttribute(UmapParams_t *params, UmapParamAttr attr, ...);
cumlError_t cumlSpUmapFit(cumlHandle_t handle, float *X,
                          int num_samples, int num_dims,
                          float *embeddings, UmapParams_t *params);
cumlError_t cumlSpUmapFitSupervised(cumlHandle_t handle, float *X, float *y,
                                    int num_samples, int num_dims,
                                    float *embeddings, UmapParams_t *params);
```

**Option 2:**
```C
cumlError_t cumlSpUmapFitDefaultParams(cumlHandle_t handle, float *X,
                          int num_samples, int num_dims,
                          float *embeddings);
cumlError_t cumlSpUmapFit(cumlHandle_t handle, float *X,
                          int num_samples, int num_dims,
                          float *embeddings, int n_neighbors, int n_components, 
                          int n_epochs, float learning_rate, float min_dist, 
                          float spread, float set_op_mix_ratio, 
                          float local_connectivity,
                          float repulsion_strength, float negative_sample_rate,
                          transform_queue_size = 4.0,       /// size a double?
                          int verbose); 
/* *Supvervised(..) similar */
```

Reply:
I'm general partial to a design like option 1. Since this problem isn't going to be isolated to UMAP, I'm seeing much value in having a consistence across our C-apis. What do you guys think of using this design across the other C-apis as well? If so, should we add it to our developer guide?